File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueIVsDegenerados.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Cifras às vezes têm "chaves fracas" ou "IVs fracos" - entradas específicas
 * (tudo zero, tudo 0xFF, padrões alternados) que produzem saída degenerada
 * mesmo quando a maioria das entradas se comporta bem. Testa especificamente
 * esses casos extremos, que testes com entrada aleatória raramente cobrem.
 */
public class AtaqueIVsDegenerados implements AtaqueInterface {
    @Override
    public String nome() {
        return "IVs degenerados (zero, 0xFF, alternado)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        int tamanhoIv = alvo.tamanhoIv();
        int tamanhoBloco = 32;

        byte[] primeiro = alvo.gerarKeystreamBruto(chave, new byte[tamanhoIv], "enc", tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        Map<String, byte[]> casos = new LinkedHashMap<>();
        casos.put("zero", new byte[tamanhoIv]);
        byte[] ff = new byte[tamanhoIv];
        Arrays.fill(ff, (byte) 0xFF);
        casos.put("0xFF", ff);
        byte[] aa = new byte[tamanhoIv];
        Arrays.fill(aa, (byte) 0xAA);
        casos.put("alternado 0xAA", aa);
        byte[] c55 = new byte[tamanhoIv];
        Arrays.fill(c55, (byte) 0x55);
        casos.put("alternado 0x55", c55);
        byte[] crescente = new byte[tamanhoIv];
        for (int i = 0; i < tamanhoIv; i++) {
            crescente[i] = (byte) i;
        }
        casos.put("crescente", crescente);

        Map<String, Map<String, Object>> problemas = new LinkedHashMap<>();
        for (Map.Entry<String, byte[]> e : casos.entrySet()) {
            byte[] ks = alvo.gerarKeystreamBruto(chave, e.getValue(), "enc", tamanhoBloco);

            int metade = tamanhoBloco / 2;
            boolean periodico = Arrays.equals(
                    Arrays.copyOfRange(ks, 0, metade),
                    Arrays.copyOfRange(ks, metade, metade + metade));

            Set<Integer> unicos = new HashSet<>();
            for (byte b : ks) {
                unicos.add(b & 0xFF);
            }
            int bytesUnicos = unicos.size();
            boolean poucaVariedade = bytesUnicos < tamanhoBloco * 0.5;

            if (periodico || poucaVariedade) {
                Map<String, Object> info = new HashMap<>();
                info.put("periodico", periodico);
                info.put("bytesUnicos", bytesUnicos);
                problemas.put(e.getKey(), info);
            }
        }

        if (!problemas.isEmpty()) {
            List<String> detalhes = new ArrayList<>();
            for (Map.Entry<String, Map<String, Object>> e : problemas.entrySet()) {
                Map<String, Object> info = e.getValue();
                detalhes.add(e.getKey() + " (periódico="
                        + (((Boolean) info.get("periodico")) ? "sim" : "não")
                        + ", bytes únicos=" + info.get("bytesUnicos") + "/" + tamanhoBloco + ")");
            }
            return new ResultadoAtaque(
                    nome(), true, Severidade.ALTA, String.join("; ", detalhes), new HashMap<>(problemas));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "Nenhum dos " + casos.size() + " IVs degenerados testados produziu saída anômala");
    }
}

File: packages/rust/src/seguranca/ataques/ataque_ivs_degenerados.rs
//! Cifras às vezes têm "chaves fracas" ou "IVs fracos" - entradas específicas
//! (tudo zero, tudo 0xFF, padrões alternados) que produzem saída degenerada
//! mesmo quando a maioria das entradas se comporta bem. Testa especificamente
//! esses casos extremos, que testes com entrada aleatória raramente cobrem.

use std::collections::HashSet;

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};

pub struct AtaqueIVsDegenerados;

impl Ataque for AtaqueIVsDegenerados {
    fn nome(&self) -> String {
        "IVs degenerados (zero, 0xFF, alternado)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();
        let tamanho_iv = alvo.tamanho_iv();
        let tamanho_bloco = 32usize;

        let _primeiro = alvo.gerar_keystream_bruto(
            chave.as_bytes(),
            &vec![0u8; tamanho_iv],
            "enc",
            tamanho_bloco,
        );

        let casos: Vec<(&str, Vec<u8>)> = vec![
            ("zero", vec![0x00u8; tamanho_iv]),
            ("0xFF", vec![0xffu8; tamanho_iv]),
            ("alternado 0xAA", vec![0xaau8; tamanho_iv]),
            ("alternado 0x55", vec![0x55u8; tamanho_iv]),
            (
                "crescente",
                (0..tamanho_iv).map(|i| i as u8).collect(),
            ),
        ];

        let mut problemas: Vec<String> = Vec::new();
        for (nome, iv) in &casos {
            let ks = alvo.gerar_keystream_bruto(chave.as_bytes(), iv, "enc", tamanho_bloco);

            let metade = tamanho_bloco / 2;
            let periodico = ks[0..metade] == ks[metade..metade + metade];

            let bytes_unicos = ks.iter().copied().collect::<HashSet<u8>>().len();
            let pouca_variedade = (bytes_unicos as f64) < tamanho_bloco as f64 * 0.5;

            if periodico || pouca_variedade {
                problemas.push(format!(
                    "{} (periódico={}, bytes únicos={}/{})",
                    nome,
                    if periodico { "sim" } else { "não" },
                    bytes_unicos,
                    tamanho_bloco
                ));
            }
        }

        if !problemas.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Alta,
                &problemas.join("; "),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Nenhum dos {} IVs degenerados testados produziu saída anômala",
                casos.len()
            ),
        ))
    }
}
